keps/sig-node/0000-apparmor-ga/README.md

```diff
@@ -0,0 +1,685 @@
+| 5) Using localhost profile with invalid name                                                       | Pod **not** created. | N/A                                                                                                                                                   |
+
+Scenario 2 is the expected behavior of using AppArmor and it is included here
+for completeness.
+
+Scenario 5 represents the case of failing the existing validation, which is
+defined at [Pod API](#pod-api).
+
+### Version Skew Strategy
+
+Because the API is currently represented as (mutable) annotations, care must be
+taken for migrating to the API fields. The cases to consider are: pod create,
+pod update, PSP create, PSP update.
+
+All API skew is resolved in the API server. New Kubelets will only use the
+AppArmor values specified in the fields, and ignore the annotations.
+
+#### Pod Creation
+
+If no AppArmor annotations or fields are specified, no action is necessary.
+
+If _only_ AppArmor fields are specified, add the corresponding annotations. This
+ensures that the fields are enforced even if the node version trails the API
+version (see [Version Skew Strategy](##version-skew-strategy))
+
+If _only_ AppArmor annotations are specified, copy the values into the
+corresponding fields. This ensures that existing applications continue to
+enforce AppArmor, and prevents the kubelet from needing to resolve annotations &
+fields.
+
+If both AppArmor annotations _and_ fields are specified, the values MUST match.
+This will be enforced in API validation.
+
+To raise awareness of annotation usage (in case of old automation), a warning
+mechanism will be used to highlight that support will be dropped in v1.24.
+The mechanisms being considered are audit annotations, annotations on the
+object, events, or a warning as described in [KEP
+#1693](/keps/sig-api-machinery/1693-warnings).
+
+#### PodSecurityPolicy Enforcement
+
+The PodSecurityPolicy admission controller must continue to check the PSP object
+for annotations, as well as for fields.
+
+When setting default profiles, PSP only needs to set the field. The API
+machinery will handle setting the annotation as necessary.
+
+When enforcing allowed profiles, the PSP should check BOTH the annotations &
+fields. In most cases, they should be consistent. On pod update, the AppArmor
+annotations may differ from the fields. In that case, the PSP enforcement should
+check both values as the effective value depends on the node version running the
+pod.
+
+#### Pod Update
+
+The AppArmor fields on a pod are immutable.
+
+The behavior on annotation update is currently ill-defined: the annotation
+update is allowed, but the new value will not be used until the container is
+restarted. There is no way to tell (from the API) what value a container is
+using.
```

Comment:
Unlike seccomp, the apparmor annotations actually are immutable: https://github.com/kubernetes/kubernetes/blob/b46612a74224b0871a97dae819f5fb3a1763d0b9/pkg/apis/core/validation/validation.go#L177-L182

Reply:
Ah thank you for the hint, I changed the paragraph to reflect that.